src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspace.cs

```diff
@@ -114,5 +114,19 @@ public PortableExecutableReference CreatePortableExecutableReference(string file
         }
 
         internal abstract string TryGetRuleSetPathForProject(ProjectId projectId);
+
+        /// <summary>
+        /// Attempts to sync a named type name and namespace based on file path. PreviousFilePath is 
+        /// used to determine how the policy might be applied, assuming that matching types with previous
+        /// indicates a user wants the type to match with the new location or name.
+        /// May fail if:
+        /// - The appropriate type for naming policy cannot be found
+        /// - The type is partial and spread across multiple files
+        /// - The solution changes during the operation
+        /// </summary>
+        /// <returns> 
+        /// Returns false if the symbol update failed, true if it succeeded
+        /// </returns>
+        public abstract Task<bool> TryUpdateNamedTypeToMatchFilePath(DocumentId documentId, string previousFilePath, CancellationToken cancellation = CancellationToken.None);
```

Comment:
Wondering if we really should have the caller pass the original Document and not just the file path. For example, if we wanted to add a feature of "update the file headers to match what's in .editorconfig now", you potentially need to know what the old document was to check the old .editorconfig rules being applied. If we passed the document that's easy, otherwise with this it's really tricky.

So maybe the signature should be `(DocumentId newDocumentId, Document oldDocument, CancellationToken cancellationToken….)`? And maybe yes all for now we're doing is asking for the file path but down the road we have extra flexibility?

Reply:
That requires the caller to hold onto the old document. If @jmarolf  doesn't have issue from project system side then I agree it affords us more flexibility in the future.

Reply:
I'd say hold onto it.

Reply:
If we go to an old document paradigm, we shouldn't need any path information right? We can determine previous file path and name from the old document. That should clean up the API surface